### PR TITLE
Move generateSpend() into AbstractConserveAmount (revised)

### PR DESCRIPTION
**Pull request by rnicoll**
Move generateSpend() into AbstractConserveAmount and make it generic enough to use for other
similar contracts to cash, rather than just cash. Also removes duplicated logic in gathering
coins.

### CODE REVIEW

File: finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt [hunk 1]
```diff
@@ -14,37 +16,116 @@ import java.util.*
...
+                        }
+            } else states
+
+            for (state in gathered) tx.addInputState(state)
+            for (state in outputs) tx.addOutputState(state)
+
+            // What if we already have a move command with the right keys? Filter it out here or in platform code?
+            tx.addCommand(Cash().generateMoveCommand(), keysUsed)
+
+            // update Vault
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: This is dead code that should be removed, right?

File: finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt [hunk 2]
```diff
@@ -14,37 +16,116 @@ import java.util.*
...
-        val gathered = arrayListOf<StateAndRef<S>>()
-        var gatheredAmount = Amount(0, amount.token)
-        for (c in acceptableCoins) {
-            if (gatheredAmount >= amount) break
-            gathered.add(c)
-            gatheredAmount += Amount(c.state.data.amount.quantity, amount.token)
+        fun <S : FungibleAsset<T>, T: Any> generateSpend(tx: TransactionBuilder,
+                          amount: Amount<T>,
+                          to: CompositeKey,
+                          acceptableCoins: List<StateAndRef<S>>,
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: acceptableCoins -> acceptableStates

File: finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt [hunk 3]
```diff
@@ -14,37 +16,116 @@ import java.util.*
...
-    private fun gatherCoins(acceptableCoins: Collection<StateAndRef<S>>,
-                            amount: Amount<T>): Pair<ArrayList<StateAndRef<S>>, Amount<T>> {
-        require(amount.quantity > 0) { "Cannot gather zero coins" }
-        val gathered = arrayListOf<StateAndRef<S>>()
-        var gatheredAmount = Amount(0, amount.token)
-        for (c in acceptableCoins) {
-            if (gatheredAmount >= amount) break
-            gathered.add(c)
-            gatheredAmount += Amount(c.state.data.amount.quantity, amount.token)
+        fun <S : FungibleAsset<T>, T: Any> generateSpend(tx: TransactionBuilder,
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: api docs!

Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: Should be `@JvmStatic` I think, for Java users.

File: core/src/main/kotlin/net/corda/core/contracts/FungibleAsset.kt [hunk 4]
```diff
@@ -20,6 +24,177 @@ class InsufficientBalanceException(val amountMissing: Amount<*>) : FlowException
...
+                gathered.add(c)
+                gatheredAmount += Amount(c.state.data.amount.quantity, amount.token)
+            }
+
+            if (gatheredAmount < amount) {
+                log.trace { "Insufficient balance: requested $amount, available $gatheredAmount" }
+                throw InsufficientBalanceException(amount - gatheredAmount)
+            }
+
+            log.trace("Gathered coins: requested $amount, available $gatheredAmount, change: ${gatheredAmount - amount}")
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/contracts/FungibleAsset.kt]: make this lazy also

File: core/src/main/kotlin/net/corda/core/contracts/FungibleAsset.kt [hunk 5]
```diff
@@ -20,6 +24,177 @@ class InsufficientBalanceException(val amountMissing: Amount<*>) : FlowException
...
+            val gathered = arrayListOf<StateAndRef<S>>()
+            var gatheredAmount = Amount(0, amount.token)
+            for (c in acceptableCoins) {
+                if (gatheredAmount >= amount) break
+                gathered.add(c)
+                gatheredAmount += Amount(c.state.data.amount.quantity, amount.token)
+            }
+
+            if (gatheredAmount < amount) {
+                log.trace { "Insufficient balance: requested $amount, available $gatheredAmount" }
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/contracts/FungibleAsset.kt]: I'm not a fan of removing trace information (which is typically for aiding internal debugging - NOT for end developer).
The original trace displayed the balance to aid soft locking debugging:
`log.trace("Insufficient balance: requested $amount, available $gatheredAmount (total balance ${cashBalances[amount.token]})")`

File: node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt [hunk 6]
```diff
@@ -450,7 +450,7 @@ class NodeVaultService(private val services: ServiceHub, dataSourceProperties: P
         // Retrieve unspent and unlocked cash states that meet our spending criteria.
         val acceptableCoins = unconsumedStatesForSpending<Cash.State>(amount, onlyFromParties, tx.notary, tx.lockId)
         return AbstractConserveAmount.generateSpend(tx, amount, to, acceptableCoins,
-                { state, amount, owner -> deriveState(state, amount, owner) }, onlyFromParties)
```
Comment (josecoll) [inline node/src/main/kotlin/net/corda/node/services/vault/NodeVaultService.kt]: why have you removed `onlyFromParties` ?

File: finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt [hunk 7]
```diff
@@ -149,41 +26,14 @@ abstract class AbstractConserveAmount<S : FungibleAsset<T>, C : CommandData, T :
      * the responsibility of the caller to check that they do not attempt to exit funds held by others.
      * @return the public key of the assets issuer, who must sign the transaction for it to be valid.
      */
+    @Deprecated("Use FungibleAsset.generateExit() instead")
```
Comment (josecoll) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: why not remove completely?

Comment (rnicoll) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: Backwards compatibility for anyone using the existing function. I'll remove it for M12 or thereabouts.

**Comment by josecoll** [comment]
In general, why don't we align the three lifecycle command implementations into a single FungibleAssetImpl source file. Currently we have:
generateExit in AbstractConserveAmount.kt (remove this file completely - does `verify()` belong here on its own?)
generateIssue in Cash.kt and CommodityContract.kt (duplicate code)
generateSpend in NodeVaultService.kt

**Comment by rnicoll** [comment]
`generateIssue` is tricky because the required inputs vary depending on the contract. I've created a standardised function in `FungibleAsset` that takes in a pre-assembled state, and left convenience methods on each contract that take in the correct inputs.

**Comment by rnicoll** [comment]
Reworked the trace calls so that if the vault service is available, and it's a currency type, it will add cash available.

File: core/src/main/kotlin/net/corda/core/contracts/FungibleAsset.kt [hunk 8]
```diff
@@ -141,11 +145,17 @@ interface FungibleAsset<T : Any> : OwnableState {
             }
 
             if (gatheredAmount < amount) {
-                log.trace { "Insufficient balance: requested $amount, available $gatheredAmount" }
+                log.trace {
+                    val token = amount.token
+                    if (token is Currency && vaultService != null)
```
Comment (josecoll) [inline core/src/main/kotlin/net/corda/core/contracts/FungibleAsset.kt]: Apologies - I did not realise you need a reference to the VaultService to display the balance.
This now looks rather ugly and unnecessary. Please revert.

**Comment by rnicoll** [comment]
Rebased and addressed comments

**Comment by rnicoll** [comment]
Re-re-rebased; could I get a review before it needs rebasing again please? :)

File: finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt (hunk 7 above)
Comment (josecoll) [inline finance/src/main/kotlin/net/corda/contracts/clause/AbstractConserveAmount.kt]: OK - good stuff.
I've been using the fully version of the annotation in the Vault API:
eg. 
`@Deprecated("This function will be removed in a future milestone", ReplaceWith("trackBy(QueryCriteria())"))>`

**Comment by mikehearn** [comment]
Sorry, I messed up. FungibleAsset is an interface. The given move has two problems:

1) What does a companion object inside an interface look like to Java? I have no clue and if I don't, I bet lots of people don't. Suggesting we move it into FungibleAsset wasn't well thought through. Perhaps it should just be a static method in a singleton object like "SpendCalculator" or something.

2) Interfaces should be easy to read for developers, but this one now has a huge pile of implementation code in it, which sits in the core module. The way this code works should be internal, not dropped in the middle of an interface definition. If we move the implementation, perhaps into a separate file, this would no longer be an issue. 

Apologies for being a pain.

**Comment by rnicoll** [comment]
It appears in Java as `FungibleAsset.Companion`, which doesn't seem too terrible. Agree on the interface getting too cluttered, going to try putting it in `OnLedgerAsset` (the base contract), which should be neater.

**Comment by rnicoll** [comment]
@mikehearn better?

File: docs/source/changelog.rst [hunk 9]
```diff
@@ -14,6 +14,9 @@ UNRELEASED
 
     * Starting a flow no longer enables progress tracking by default. To enable it, you must now invoke your flow using one of the new ``CordaRPCOps.startTrackedFlow`` functions. ``FlowHandle`` is now an interface, and its ``progress: Observable`` field has been moved to the ``FlowProgressHandle`` child interface. Hence developers no longer need to invoke ``notUsed`` on their flows' unwanted progress-tracking observables.
 
+    * Moved ``generateSpend`` and ``generateExit`` functions into ``FungibleAsset`` from the vault and
```
Comment (mikehearn) [inline docs/source/changelog.rst]: This needs an update now.

File: finance/src/main/kotlin/net/corda/contracts/asset/OnLedgerAsset.kt [hunk 10]
```diff
@@ -24,8 +27,192 @@ import java.security.PublicKey
...
+         * @param acceptableStates a list of acceptable input states to use.
+         * @param deriveState a function to derive an output state based on an input state, amount for the output
+         * and public key to pay to.
+         * @return A [Pair] of the same transaction builder passed in as [tx], and the list of keys that need to sign
+         *         the resulting transaction for it to be valid.
+         * @throws InsufficientBalanceException when a cash spending transaction fails because
+         *         there is insufficient quantity for a given currency (and optionally set of Issuer Parties).
+         */
+        @Throws(InsufficientBalanceException::class)
+        fun <S : FungibleAsset<T>, T: Any> generateSpend(tx: TransactionBuilder,
```
Comment (mikehearn) [inline finance/src/main/kotlin/net/corda/contracts/asset/OnLedgerAsset.kt]: `@JvmStatic` on these public companion methods so they're more natural from java.

**Pull request merged**